pkg/kubelet/kuberuntime/kuberuntime_manager_test.go

```diff
@@ -955,6 +1095,306 @@ func TestComputePodActions(t *testing.T) {
 	}
 }
 
+func makeBasePodAndStatusWithSidecar() (*v1.Pod, *kubecontainer.PodStatus) {
+	pod, status := makeBasePodAndStatus()
+	sidecarType := v1.LifecycleTypeSidecar
+	pod.Spec.Containers[1].Lifecycle = &v1.Lifecycle{Type: &sidecarType}
+	status.ContainerStatuses[1].Hash = kubecontainer.HashContainer(&pod.Spec.Containers[1])
+
+	return pod, status
+}
+
+func TestComputePodActionsWithSidecar(t *testing.T) {
+	_, _, m, err := createTestRuntimeManager()
+	require.NoError(t, err)
+
+	// Creating a pair reference pod and status for the test cases to refer
+	// the specific fields.
+	basePod, baseStatus := makeBasePodAndStatusWithSidecar()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SidecarLifecycle, true)()
+	for desc, test := range map[string]struct {
+		mutatePodFn    func(*v1.Pod)
+		mutateStatusFn func(*kubecontainer.PodStatus)
+		actions        podActions
+	}{
+		"Kill sidecars if all non-sidecars are terminated": {
+			mutatePodFn: func(pod *v1.Pod) {
+				pod.Spec.RestartPolicy = v1.RestartPolicyNever
+			},
+			mutateStatusFn: func(status *kubecontainer.PodStatus) {
+				for i := range status.ContainerStatuses {
+					if i == 1 {
+						continue
+					}
+					status.ContainerStatuses[i].State = kubecontainer.ContainerStateExited
+					status.ContainerStatuses[i].ExitCode = 0
+				}
+			},
+			actions: podActions{
+				KillPod:           true,
+				SandboxID:         baseStatus.SandboxStatuses[0].Id,
+				ContainersToKill:  getKillMap(basePod, baseStatus, []int{1}),
+				ContainersToStart: []int{},
+			},
+		},
+		"Kill pod if sidecar terminates with failure": {
```

Comment:
If a sidecar exits with a failure, is the intention to KillPod & recreate the sandbox? Or can we simply let the sidecar get restarted? IIUC there's no explicit mention of what we're going to do in this scenario in the spec: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/sidecarcontainers.md. But I might be missing something.

Reply:
It seems restarting sidecar container(s) would be more efficient.

Reply:
This is perhaps a weirdly named test, what this test is proving is that when restart policy == onfailure, if the sidecar exited with a non 0 code and all the other containers have exited successfully we don't restart the sidecar and do kill the pod. 
This is so we don't end up in a weird restart loop, e.g non-sidecars finish, sidecar gets sent sigterm, exits with non-zero, gets restarted, get sent sigterm etc, exits with non-zero etc, etc

Reply:
Got it, how about "Kill pod if sidecar fails, but all non-sidecars succeeded" ?

Reply:
yeah will rename 👍